pds_doi_core/util/doi_validator.py

```diff
@@ -0,0 +1,323 @@
+
+        return 1
+
+    def _check_field_lidvid_update_A(self, doi: Doi, action):
+        """ If DOI does not have a doi field, and action step is 'release' there is no DOI in sqllite database with the same lidvid and a DOI attributed.
+            if lidvid exist and doi exist, throw an exception if action is 'release'
+            if lidvid exist and doi exist, if action is not 'release', nothing to do.
+            if lidvid does not exist, nothing to do.
+        """
+
+        query_criterias = {}
+        query_criterias['lidvid'] = [doi.related_identifier]  # The database expect each field to be a list.
+
+        logger.debug(f"action,doi [{action}] {doi}")
+
+        # Query database for rows with given lidvid value.
+        columns, rows = self._database_obj.select_latest_rows(query_criterias)
+
+        # The number of times the same lidvid used is the number of rows.  It should be either 0 or 1.
+        num_times_lidvid_used = len(rows)
+
+        logger.debug(f"lidvid {query_criterias['lidvid']}, len(rows) {len(rows)}")
+
+        col_index_to_doi     = columns.index('doi')
+        col_index_to_status  = columns.index('status')
+
+        for row_index,row_value in enumerate(rows):
+            logger.debug(f"row_index,status,row_value[col_index_to_doi] {row_index,row_value[col_index_to_status],row_value[col_index_to_doi]}")
+
+        if num_times_lidvid_used > 0:
+            last_row_index       = num_times_lidvid_used - 1
+            col_index_to_title   = columns.index('title')
+            col_index_to_lid     = columns.index('lid')
+            col_index_to_vid     = columns.index('vid')
+            logger.debug(f"action: {action} rows[last_row_index][col_index_to_doi] {rows[last_row_index][col_index_to_doi]}")
+            lidvid = rows[last_row_index][col_index_to_lid] + '::' + rows[last_row_index][col_index_to_vid]
+            if rows[last_row_index][col_index_to_doi]:
+              # Field 'lidvid' has been used and 'doi' field has been used and action is 'reserve' throw an exception.
+              # Essentially does not allow a 'reserve' action on the same lidvid/doi pair.
+              if action == 'reserve':
+                logger.error(f"There is already a DOI {rows[last_row_index][col_index_to_doi]} submitted for this lidvid {doi.related_identifier} (status={rows[last_row_index][col_index_to_status]}). You cannot submit a new DOI for the same lidvid.")
+                logger.debug(f"CHECK_FIELD_LIDVID_EXIST_TRUE_DOI_EXIST_TRUE_ACTION_TRUE_RESERVE: lidvid: {lidvid} doi: {rows[last_row_index][col_index_to_doi]} action: {action}")
+                raise IllegalDOIActionException(f"There is already a DOI {rows[last_row_index][col_index_to_doi]} submitted for this lidvid {doi.related_identifier} (status={rows[last_row_index][col_index_to_status]}). You cannot submit a new DOI for the same lidvid.")
+              else:
```

Comment:
Once the code works, may be you can remove all these debug messages which are make the reading of the code less comfortable.

Reply:
Will remove the debug statements once done testing.